api/src/main/java/org/pmiops/workbench/api/ConceptSetsController.java

```diff
@@ -0,0 +1,249 @@
+      ConceptSet conceptSet) {
+    Workspace workspace = workspaceService.getWorkspaceEnforceAccessLevelAndSetCdrVersion(
+        workspaceNamespace, workspaceId, WorkspaceAccessLevel.WRITER);
+    org.pmiops.workbench.db.model.ConceptSet dbConceptSet = FROM_CLIENT_CONCEPT_SET.apply(conceptSet);
+    Timestamp now = new Timestamp(clock.instant().toEpochMilli());
+    dbConceptSet.setCreator(userProvider.get());
+    dbConceptSet.setWorkspaceId(workspace.getWorkspaceId());
+    dbConceptSet.setCreationTime(now);
+    dbConceptSet.setLastModifiedTime(now);
+    dbConceptSet.setVersion(1);
+    try {
+      dbConceptSet = conceptSetDao.save(dbConceptSet);
+      // TODO: add recent resource entry for concept sets?
+    } catch (DataIntegrityViolationException e) {
+      throw new BadRequestException(String.format(
+          "Concept set \"/%s/%s/%d\" already exists.",
+          workspaceNamespace, workspaceId, dbConceptSet.getName()));
+    }
+    return ResponseEntity.ok(toClientConceptSet(dbConceptSet));
+  }
+
+  private ConceptSet toClientConceptSet(org.pmiops.workbench.db.model.ConceptSet conceptSet) {
+    ConceptSet result = TO_CLIENT_CONCEPT_SET.apply(conceptSet);
+    if (!conceptSet.getConceptIds().isEmpty()) {
+      result.setConcepts(ImmutableList.copyOf(conceptDao.findAll(conceptSet.getConceptIds())).stream().map(
+            ConceptsController.TO_CLIENT_CONCEPT).sorted(CONCEPT_NAME_ORDERING).collect(Collectors.toList()));
+    }
+    return result;
+  }
+
+  @Override
+  public ResponseEntity<EmptyResponse> deleteConceptSet(String workspaceNamespace,
+      String workspaceId, Long conceptSetId) {
+    org.pmiops.workbench.db.model.ConceptSet conceptSet =
+        getDbConceptSet(workspaceNamespace, workspaceId, conceptSetId, WorkspaceAccessLevel.WRITER);
+    conceptSetDao.delete(conceptSet.getConceptSetId());
+    return ResponseEntity.ok(new EmptyResponse());
+  }
+
+  @Override
+  public ResponseEntity<ConceptSet> getConceptSet(String workspaceNamespace, String workspaceId,
+      Long conceptSetId) {
+    org.pmiops.workbench.db.model.ConceptSet conceptSet =
+        getDbConceptSet(workspaceNamespace, workspaceId, conceptSetId, WorkspaceAccessLevel.READER);
+    return ResponseEntity.ok(toClientConceptSet(conceptSet));
+  }
+
+  @Override
+  public ResponseEntity<ConceptSetListResponse> getConceptSetsInWorkspace(String workspaceNamespace,
+      String workspaceId) {
+    workspaceService.getWorkspaceEnforceAccessLevelAndSetCdrVersion(workspaceNamespace, workspaceId,
+        WorkspaceAccessLevel.READER);
+
+    Workspace workspace = workspaceService.getRequiredWithConceptSets(workspaceNamespace, workspaceId);
+    ConceptSetListResponse response = new ConceptSetListResponse();
+    Set<org.pmiops.workbench.db.model.ConceptSet> conceptSets = workspace.getConceptSets();
+    if (conceptSets != null) {
+      Set<Long> conceptIds = Sets.newHashSet();
```

Comment:
Is the relationship here that each conceptSet has many conceptIds?

Reply:
That is correct. We may want to place a limit on the number of concepts that can be in a single set; I will add that before I merge I think.